File: 3635_Earliest_Finish_Time_for_Land_and_Water_Rides_II/solution.c
#define mn(a, b) ((a) < (b) ? (a) : (b))
#define mx(a, b) ((a) > (b) ? (a) : (b))

int f(int* a, int n, int* b, int* c, int m, int* d) {
    int x = INT_MAX, y = INT_MAX;
    for (int i = 0; i < n; i++) x = mn(x, a[i] + b[i]);
    for (int i = 0; i < m; i++) y = mn(y, mx(c[i], x) + d[i]);
    return y;
}

int earliestFinishTime(int* a, int n, int* b, int bn, int* c, int m, int* d, int dn) {
    return mn(f(a, n, b, c, m, d), f(c, m, d, a, n, b));
}

File: 3635_Earliest_Finish_Time_for_Land_and_Water_Rides_II/solution.py
class Solution:
    def earliestFinishTime(self, a, b, c, d):
        def f(s1, t1, s2, t2):
            x = min(s1[i] + t1[i] for i in range(len(s1)))
            y = min(max(s2[i], x) + t2[i] for i in range(len(s2)))
            return y
        return min(f(a, b, c, d), f(c, d, a, b))
